### PR TITLE
fix(security): block SSRF via provider baseUrl

### PR DESCRIPTION
**Pull request by steebchen**
## Summary

Fixes the tenant-controlled provider `baseUrl` SSRF reported in **GHSA-8xr7-xg68-p8c5** and **GHSA-3r2h-6gmc-w7c2** (CWE-918 / CWE-862, CVSS 7.x High).

Any authenticated org member (including the lowest `developer` role) could register a custom/BYOK provider with a `baseUrl` pointing at an internal address (`127.0.0.1`, `169.254.169.254`, RFC-1918, internal hostnames, …) — or a plain `http://` host. The URL was stored without validation and used verbatim as the gateway's `fetch()` target, enabling full-read SSRF against internal services — both at registration time (response reflected in the 400) and at inference time (raw body returned).

The repo already shipped an SSRF guard (`isPrivateOrReservedIp` / `assertSafeWebhookUrl`) used on the webhook path; it was simply never wired onto the provider/gateway path.

## Behaviour / the flag

The provider base URL guard is **enforced by default in every deployment**:

- must use **https** (plain `http://` is rejected, including public hosts),
- must not resolve to a private / loopback / link-local / metadata / internal host.

Self-hosted operators who intentionally point a provider at an internal or http-only model server (e.g. a local Ollama) can opt out by setting `ALLOW_INSECURE_PROVIDER_URLS=true`. This is documented in `.env.example` and should be left unset on any multi-tenant / cloud deployment.

## Changes

- **`packages/shared`**: add `assertSafeProviderBaseUrl` (sync: https + blocked-host + private/reserved IP-literal check), `assertSafeProviderUrl` (node-only, async — additionally DNS-resolves the host and re-checks every resolved IP to defeat DNS rebinding), and `isProviderUrlGuardEnabled()` (the single source of truth for the `ALLOW_INSECURE_PROVIDER_URLS` opt-out). The DNS-resolving helper lives in a new `@llmgateway/shared/url-safety-node` entrypoint so `node:dns` never leaks into the browser barrel.
- **Registration** (`apps/api/.../keys-provider.ts`): validate `baseUrl` with the DNS-resolving guard before it is stored — covers custom providers (whose key validation is skipped) and non-custom providers (whose validation fetch reflects the body).
- **`getProviderEndpoint`**: validate the tenant `baseUrl` at the single point where it becomes the request URL — the universal chokepoint for every fetch path.
- **Inference** (`chat.ts` main path + `resolve-provider-context.ts` retry path): re-run the DNS-resolving guard before the outbound `fetch` to also defeat rebinding and protect any URL stored before this fix.
- **Authorization** (`projects.ts`): restrict `PATCH /projects/{id}` `mode` changes to owners/admins — closes the `developer`-role bypass that let a member flip the project into the BYOK mode the inference SSRF path needs.
- **Test env**: the guard is relaxed in `vitest/test-database-setup.ts` so the existing localhost mock-server fixtures keep working.

## Tests

- New `packages/shared/src/url-safety.spec.ts` covering `isPrivateOrReservedIp`, `assertSafeProviderBaseUrl` (https-only, loopback, RFC-1918, link-local/metadata, IPv6, internal hostnames), and `isProviderUrlGuardEnabled`.
- New cases in `get-provider-endpoint.spec.ts` asserting internal/http base URLs are rejected by default and allowed when opted out.
- `pnpm build` and `pnpm format` pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added SSRF protection for custom provider base URLs (HTTPS-only, no private/internal targets), with an opt-in override for trusted self-hosted setups.
  * Introduced a shared Node URL-safety utility and exposed it for use across services.
* **Bug Fixes**
  * Strengthened project update/create authorization so only owners/admins can change sensitive end-user settings or non-default modes.
* **Documentation**
  * Updated the environment configuration example to describe the default SSRF validation and override.
* **Tests**
  * Added/extended e2e coverage for SSRF blocking and redirect handling.

### CODE REVIEW

File: apps/api/src/routes/keys-provider.ts [hunk 1]
```diff
@@ -212,6 +213,22 @@ keysProvider.openapi(create, async (c) => {
 		});
 	}
 
+	// SSRF guard: reject base URLs that resolve to internal/reserved addresses
+	// before they are stored or used as an outbound fetch target. No-op unless
+	// the hosted provider URL guard is enabled.
+	if (baseUrl) {
+		try {
+			await assertSafeProviderUrl(baseUrl);
```
[the remark this answers is not in the document]
Comment (steebchen) [reply apps/api/src/routes/keys-provider.ts:221]: Addressed in c7d896038. `redirect: "error"` is now set on **every** authenticated provider fetch, not just chat:

- `apps/gateway/src/embeddings/embeddings.ts`
- `apps/gateway/src/speech/speech.ts`
- `apps/gateway/src/moderations/moderations.ts`
- `packages/actions/src/validate-provider-key.ts` (registration-time validation)
- plus the two chat paths

So a public baseUrl that 302s onward to localhost/metadata on any of those endpoints is refused before the redirect is followed. Added an e2e test in `chat-custom-provider.e2e.ts` asserting that a custom provider which 302s to an internal endpoint does not leak the internal response to the caller.

File: packages/shared/src/url-safety-node.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+export async function assertSafeProviderUrl(rawUrl: string): Promise<void> {
+	if (!isProviderUrlGuardEnabled()) {
+		return;
+	}
+
+	const url = assertSafeProviderBaseUrl(rawUrl);
+
+	const resolved = await lookup(url.hostname, { all: true });
+	for (const { address } of resolved) {
+		if (isPrivateOrReservedIp(address)) {
```
[the remark this answers is not in the document]
Comment (steebchen) [reply packages/shared/src/url-safety-node.ts:36]: Addressed in 1e579e671. `isPrivateOrReservedIp` now also rejects the IANA special-use IPv4 ranges:

- `192.0.0.0/24` (IETF protocol assignments)
- `192.0.2.0/24` (TEST-NET-1), `198.51.100.0/24` (TEST-NET-2), `203.0.113.0/24` (TEST-NET-3)
- `192.88.99.0/24` (6to4 relay anycast)
- `198.18.0.0/15` (benchmarking — your example)

`240.0.0.0/4` was already covered by the existing `>= 224` reserved check. Added unit tests including boundary cases just below/above `198.18.0.0/15`.

File: packages/shared/src/url-safety-node.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * so a public host cannot 3xx the request onward to an internal address.
+ */
+export async function assertSafeProviderUrl(rawUrl: string): Promise<void> {
+	if (!isProviderUrlGuardEnabled()) {
+		return;
+	}
+
+	const url = assertSafeProviderBaseUrl(rawUrl);
+
+	const resolved = await lookup(url.hostname, { all: true });
```
[the remark this answers is not in the document]
Comment (steebchen) [reply packages/shared/src/url-safety-node.ts:34]: Intentional — the runtime DNS re-check was dropped in favor of registration-time validation + redirect blocking, for two reasons:

1. **TLS already defeats the read-path of DNS rebinding.** baseUrl is enforced https-only at registration. For a rebind to *read* an internal service, the TLS handshake to `https://evil.com` (now resolving to the internal IP) must succeed — the internal target (the gateway's own loopback, the cloud metadata endpoint, an internal microservice) would have to present a CA-valid certificate for `evil.com`, which the attacker cannot install there. Cert validation is on (we never set `NODE_TLS_REJECT_UNAUTHORIZED=0` / `rejectUnauthorized: false`), so a rebind yields a TCP connect at most, not a body read.
2. **The redirect vector — the part https does *not* cover — is now closed** with `redirect: "error"` on all provider fetches (c7d896038).

A connect-time IP-pinning dispatcher was prototyped but it pulls in an `undici` dependency with version-skew risk against Node's bundled undici (7.18.2), to cover a residual that TLS already covers. Happy to revisit if you can point at a concrete read-path that survives cert validation.

File: apps/gateway/src/chat/chat.ts [hunk 4]
```diff
@@ -4800,6 +4801,14 @@ chat.openapi(completions, async (c) => {
 			usedInternalModel,
 		);
 
+		// SSRF guard: a tenant-supplied provider base URL must not resolve to an
+		// internal/reserved address. Re-checked here (in addition to registration)
+		// to also defeat DNS rebinding and any URL stored before the guard
+		// existed. No-op unless the hosted guard is enabled.
+		if (providerKey?.baseUrl) {
+			await assertSafeProviderUrl(providerKey.baseUrl);
+		}
```
[the remark this answers is not in the document]
Comment (steebchen) [reply apps/gateway/src/chat/chat.ts]: The runtime `assertSafeProviderUrl` call this referenced was removed. Validation now runs once at registration (`POST /keys/provider`), which already wraps it in an explicit try/catch returning a clean **400** (`apps/api/src/routes/keys-provider.ts`). There's no longer a provider-setup catch turning a URL-safety rejection into a 500.

File: apps/api/src/routes/keys-provider.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (steebchen) [reply apps/api/src/routes/keys-provider.ts:221]: Addressed in fc855ce93. `redirect: "error"` is now set on all four video upstream fetches:

- `fetchUpstreamJson()` (create/status API) — `videos.ts:2654`
- `/v1/files/retrieve` (MiniMax download-URL lookup) — `videos.ts:2543`
- the content/download fetch — `videos.ts:2574`
- `streamVideoFromUrl()` (signed/direct content) — `videos.ts:2421`

So a public BYOK video baseUrl can no longer 302 a create/status/content request onward to localhost/metadata and have the followed body streamed back.

I swept the rest of the gateway's `fetch` sites while here; the remaining ones are not tenant-controlled (`openai-content-filter` uses the hardcoded OpenAI URL, `mcp.ts` calls the gateway's own URL, `gcp-token`/`vertex-openai-token` hit Google's OAuth endpoint from operator service-account creds). One separate vector that's outside this advisory's scope: `convert-images-to-base64.ts` fetches user-supplied `image_url`s — that's a distinct image-SSRF concern, happy to file/fix separately if you want.
